File: Python/2.Add Two Numbers.py
#把两个链表数相加
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        L1,L2,L3 = [],[],[]
        while l1 != None:
            L1.append(l1.val)
            l1 = l1.next
        while l2 != None:
            L2.append(l2.val)
            l2 = l2.next
        
        if len(L1)>=len(L2):
            L3 = L1
        else:
            L3 = L2
            L2 = L1
        flag = 0
        for i in range(len(L2)):
            tmp = L3[i]+L2[i]+flag
            L3[i] = tmp%10
            flag = int(tmp/10)

        if len(L3)!=len(L2):
            for i in range(len(L2),len(L3)):
                 tmp = L3[i]+flag
                 L3[i] = tmp%10
                 flag = int(tmp/10)
        if flag!=0:
            L3.append(flag)
        return L3

l1 = ListNode(2)
l2 = ListNode(4)
l3 = ListNode(3)
l1.next = l2
l2.next = l3

l4 = ListNode(5)
l5 = ListNode(6)
l6 = ListNode(4)
l4.next = l5
l5.next = l6

l7 = ListNode(5)
l8 = ListNode(5)

l9 = ListNode(9)
l10 = ListNode(9)
l9.next = l10
l11 = ListNode(1)

s = Solution()
L = s.addTwoNumbers(l1,l4)
for i in range(len(L)-1):
     print('%d -> ' % L[i],end="")
print(L[-1])

File: C++/Math/2.Add Two Numbers.cpp
/*
两个链表表示两个整数，返回两个数的和的链表。
*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode nHead(0);
        ListNode *cur = &nHead;
        int flag=0,temp;
        while(l1 || l2 || flag!=0)
        {
        	temp = flag;
        	if(l1)	temp += l1->val,l1=l1->next;
        	if(l2)	temp += l2->val,l2=l2->next;
        	flag = temp/10;
        	temp = temp%10;
        	ListNode *tmp = new ListNode(temp);
        	cur->next = tmp;
        	cur=tmp; 
        }
        return nHead.next;
    }
};
int main()
{
	Solution s;
	ListNode head(2);
	ListNode t1(4);
	ListNode t2(3);
	ListNode t3(5);
	ListNode t4(6);
	ListNode t5(4);
	head.next = &t1;
	t1.next = &t2;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.addTwoNumbers(&head,&t3);
 	while(h)
 	{
 		cout<<h->val;
 		h = h->next;
 	}
}

